### PR TITLE
feat(amplify-category-function): Support Scheduling in Lambda functions

### PR DESCRIPTION
**Pull request by akshbhu**
*Description of changes:*
* added feature in function to support scheduling cloudwatch event rule
* added integration test for the scheduling

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yuth** [comment]
# [Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599?src=pr&el=h1) Report
> Merging [#3599](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599?src=pr&el=desc) into [master](https://codecov.io/gh/aws-amplify/amplify-cli/commit/553186e53050cafdf27120443d176023ef4acebc&el=desc) will **increase** coverage by `0.37%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master    #3599      +/-   ##
==========================================
+ Coverage   59.53%   59.90%   +0.37%     
==========================================
  Files         286      285       -1     
  Lines       13240    13158      -82     
  Branches     2663     2649      -14     
==========================================
  Hits         7882     7882              
+ Misses       4933     4852      -81     
+ Partials      425      424       -1     
```


| [Impacted Files](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [...kages/amplify-util-mock/src/utils/lambda/invoke.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL3V0aWxzL2xhbWJkYS9pbnZva2UudHM=) | | |
| [packages/amplify-util-mock/src/utils/ddb-utils.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL3V0aWxzL2RkYi11dGlscy50cw==) | | |
| [...kages/amplify-util-mock/src/\_\_e2e\_\_/utils/index.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL19fZTJlX18vdXRpbHMvaW5kZXgudHM=) | | |
| [...ges/amplify-util-mock/src/utils/config-override.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL3V0aWxzL2NvbmZpZy1vdmVycmlkZS50cw==) | | |
| [packages/amplify-util-mock/src/utils/git-ignore.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL3V0aWxzL2dpdC1pZ25vcmUudHM=) | | |
| [packages/amplify-util-mock/src/index.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL2luZGV4LnRz) | | |
| [packages/amplify-util-mock/src/mockAll.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL21vY2tBbGwudHM=) | | |
| [...ackages/amplify-util-mock/src/utils/lambda/load.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL3V0aWxzL2xhbWJkYS9sb2FkLnRz) | `0.00% <0.00%> (ø)` | |
| [...plify-util-mock/src/api/run-graphql-transformer.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL2FwaS9ydW4tZ3JhcGhxbC10cmFuc2Zvcm1lci50cw==) | `0.00% <0.00%> (ø)` | |
| [...l-mock/src/CFNParser/resource-processors/lambda.ts](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree#diff-cGFja2FnZXMvYW1wbGlmeS11dGlsLW1vY2svc3JjL0NGTlBhcnNlci9yZXNvdXJjZS1wcm9jZXNzb3JzL2xhbWJkYS50cw==) | `0.00% <0.00%> (ø)` | |
| ... and [3 more](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599?src=pr&el=footer). Last update [553186e...144384e](https://codecov.io/gh/aws-amplify/amplify-cli/pull/3599?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js [hunk 1]
```diff
@@ -0,0 +1,267 @@
+const path = require('path');
```
Comment (edwardfoyle) [inline packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js]: I know you made this file and other .js because that's what everything in the functions plugin is right now. But my refactor is going to change all of this to .ts so it would be great if this could also be .ts :)

File: packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+        choices: ['Update the CronJob', 'Remove the CronJob'],
+      };
+
+      const scheduleEventOperationAnswer = await inquirer.prompt([scheduleEventOperationQuestion]);
+
+      switch (scheduleEventOperationAnswer.ScheduleEventOperation) {
+        case 'Update the CronJob': {
+          // add service walkthrough to get the cron expression
+          cloudWatchRule = await cronServiceWalkthrough(context, parameters);
+          parameters.cloudwatchEvent = 'true';
```
Comment (attilah) [inline packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js]: If we store a boolean in that parameter, then why do we've NONE as the default?

File: packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    message: 'When would you like to start cron?',
+    format: ['h', ':', 'MM', ' ', 'TT'],
+  };
+  const timeAnswer = await inquirer.prompt([timeQuestion]);
+
+  const intervalQuestion = {
+    type: 'list',
+    name: 'interval',
+    message: 'Please select interval or select custom rule exp?',
+    choices: ['daily', 'weekly', 'monthly', 'yearly', 'customRule'],
```
Comment (attilah) [inline packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js]: We need to have minutes and hours as well to support like "every hour", "every 5 minutes".

File: packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+      const dateAnswer = await inquirer.prompt([dateQuestion]);
+      cloudwatchRule = makeCron(intervalAnswer.interval, dateAnswer, timeAnswer);
+      cloudwatchRule = 'cron(' + replaceAt(cloudwatchRule, cloudwatchRule.lastIndexOf('*'), '?') + ' ' + '*' + ')';
+      break;
+    }
+    case 'yearly': {
+      const dateQuestion = {
+        type: 'datetime',
+        name: 'dt',
+        message: 'select month and date to start cron?',
```
Comment (attilah) [inline packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js]: "Select"

File: packages/amplify-e2e-tests/src/__tests__/function.test.ts [hunk 5]
```diff
@@ -316,3 +316,68 @@ describe('amplify add function with additional permissions', () => {
     expect(lambdaSource.includes('GRAPHQLAPIIDOUTPUT')).toBeTruthy();
   });
 });
+
+describe.only('amplify add/update/remove function based on schedule rule', () => {
```
Comment (attilah) [inline packages/amplify-e2e-tests/src/__tests__/function.test.ts]: We need a lot more tests and test execution as well, udpate, delete flow, etc.

File: yarn.lock [hunk 6]
```diff
@@ -3793,6 +3793,16 @@ amdefine@>=0.0.4:
   resolved "https://registry.yarnpkg.com/amdefine/-/amdefine-1.0.1.tgz#4a5282ac164729e93619bcfd3ad151f817ce91f5"
   integrity sha1-SlKCrBZHKek2Gbz9OtFR+BfOkfU=
 
+amplify-frontend-javascript@2.13.2:
```
Comment (attilah) [inline yarn.lock]: Unrelated change to this PR, please remove it.

File: packages/amplify-category-function/package.json [hunk 7]
```diff
@@ -18,7 +18,9 @@
     "amplify-category-storage": "2.2.1",
     "fs-extra": "^8.1.0",
     "graphql-transformer-core": "6.14.1",
+    "cron-builder": "^0.3.0",
     "inquirer": "^7.0.3",
+    "inquirer-datepicker-prompt": "^0.4.2",
```
Comment (attilah) [inline packages/amplify-category-function/package.json]: This is an unmaintained package (2 years old), please take a look at: `inquirer-datepicker` instead.

File: packages/amplify-category-function/package.json [hunk 8]
```diff
@@ -18,7 +18,9 @@
     "amplify-category-storage": "2.2.1",
     "fs-extra": "^8.1.0",
     "graphql-transformer-core": "6.14.1",
+    "cron-builder": "^0.3.0",
```
Comment (attilah) [inline packages/amplify-category-function/package.json]: This is a 3 years old package, can we find a better actively maintained version?

File: packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs [hunk 9]
```diff
@@ -214,6 +222,35 @@
         }
         <% } %>
         <% } %>
+        <% if (props.cloudwatchEvent && props.cloudwatchEvent != "NONE") { %> 
+        ,"CloudWatchEvent": {
+            "Type" : "AWS::Events::Rule",
+            "Properties":{
+                "Description" : "Schedule rule for Lambda",
+                "ScheduleExpression" : "<%= props.cloudwatchRule %>",
```
Comment (attilah) [inline packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs]: Here the CF parameter should be referenced with `Ref` not the ejs template property

File: packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs [hunk 10]
```diff
@@ -214,6 +222,35 @@
...
+                "Targets": [{
+                    "Arn": { "Fn::GetAtt": ["LambdaFunction", "Arn"] },
+                    "Id":{ 
+                        "Ref" : "LambdaFunction"
+                    }
+                }]
+            } 
+        }
+        <% } %>          
+        <% if (props.cloudwatchEvent && props.cloudwatchEvent != "NONE") { %>
```
Comment (attilah) [inline packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs]: This if block is not required as the conditions are the same as above.

File: packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+inquirer.registerPrompt('datetime', require('inquirer-datepicker-prompt'));
+const cb = require('cron-builder');
+const cronstrue = require('cronstrue');
+const categoryName = 'function';
+
+var Days = {
+  MON: '1',
+  TUE: '2',
+  WED: '3',
+  THURS: '4',
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js]: fixed

File: packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js (hunk 2 above)
Comment (akshbhu) [reply packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js]: fixed

File: packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js (hunk 3 above)
Comment (akshbhu) [reply packages/amplify-category-function/provider-utils/awscloudformation/utils/cron.js]: Included every hour and every minute functionality

File: packages/amplify-e2e-tests/src/__tests__/function.test.ts (hunk 5 above)
Comment (akshbhu) [reply packages/amplify-e2e-tests/src/__tests__/function.test.ts]: added tests for add/update/delete

File: yarn.lock (hunk 6 above)
Comment (akshbhu) [reply yarn.lock]: removed

File: packages/amplify-category-function/package.json (hunk 7 above)
Comment (akshbhu) [reply packages/amplify-category-function/package.json]: fixed

File: packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs (hunk 10 above)
Comment (akshbhu) [reply packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs]: fixed

File: packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs [hunk 12]
```diff
@@ -2,6 +2,14 @@
     "AWSTemplateFormatVersion": "2010-09-09",
     "Description": "Lambda resource stack creation using Amplify CLI",
     "Parameters": {
+        "cloudwatchEvent": {
```
[the remark this answers is not in the document]
Comment (akshbhu) [reply packages/amplify-category-function/provider-utils/awscloudformation/cloudformation-templates/lambda-cloudformation-template.json.ejs]: fixed

**Comment by akshbhu** [comment]
closing in favor of #3714
